### PR TITLE
Using least squares for polynomial chaos

### PR DESCRIPTION
**Pull request by zachmprince**
Refs #14835

### CODE REVIEW

**Comment by dschwen** [comment]
Do you get a covariance matrix for the coefficients from this?

File: modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md [hunk 1]
```diff
@@ -42,20 +42,40 @@ The weighting functions are defined by the probability density function of the p
...
-or using numerical quadrature,
+And for using a [QuadratureSampler.md]
 
 !equation
 q_i = \frac{1}{\left\langle\Phi_i(\vec{\xi}),\Phi_i(\vec{\xi})\right\rangle}\sum_{n=1}^{N_q} w_n Q(\vec{\xi}_n)\Phi_i(\vec{\xi}_n) .
 
-
 The numerical quadrature method is typically much more efficient that than the Monte Carlo method and has the added benefit of exactly integrating the polynomial basis. However, the quadrature suffers from the curse of dimensionality. The naive approach uses a Cartesian product of one-dimensional quadratures, which results in $(\max(k^d_i) + 1)^D$ quadrature points to be sampled. Sparse grids can help mitigate the curse of dimensionality significantly.
 
+The other technique is using ordinary least-squares (OLS) regression, like in [PolynomialRegressionTrainer.md]. This is generally more accurate than integration for Monte Carlo sampling, as shown in the figure below.
```
Comment (grmnptr) [inline modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md:59]: Hmmm the accuracy might depend on the dimensionality, basis functions and the function we are trying to approximate. Unless we can cite a proof here, maybe 'generally' could be replaced by 'in many cases/instances'.

File: modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C [hunk 2]
```diff
@@ -44,53 +49,139 @@ PolynomialChaosTrainer::PolynomialChaosTrainer(const InputParameters & parameter
...
+      sig[i] = i > 0 ? _calculators[3 * i + 1]->getValue() : 1.0;
+      sum_pf[i] = _calculators[3 * i + 2]->getValue();
+    }
+
+    const Real n = getCurrentSampleSize();
+    for (const auto i : make_range(_ncoeff))
+    {
+      for (const auto j : make_range(i + 1))
+      {
+        _matrix(i, j) -= (mu[j] * sum_pf[i] + mu[i] * sum_pf[j]);
```
Comment (grmnptr) [inline modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C:169]: Do you think standardization needs to be mandatory? I would make sense, just asking if you see an application in the future where someone wants to avoid standardization.

File: modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C [hunk 3]
```diff
@@ -44,53 +49,139 @@ PolynomialChaosTrainer::PolynomialChaosTrainer(const InputParameters & parameter
...
+        _matrix(i, j) += n * mu[i] * mu[j];
+        _matrix(i, j) /= (sig[i] * sig[j]);
+        _matrix(j, i) = _matrix(i, j);
+      }
+      _rhs(i) = (_rhs(i) - mu[i] * _r_sum) / sig[i];
```
Comment (grmnptr) [inline modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C:174]: If somebody wants to use high degree polynomials, they might need some sort of a regularization. Do you have capacity for ridge now or maybe in a new PR?

File: modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C [hunk 4]
```diff
@@ -44,53 +49,139 @@ PolynomialChaosTrainer::PolynomialChaosTrainer(const InputParameters & parameter
...
 
-  if (_quad_sampler && (!_pvals.empty() || _pcols.size() != _sampler.getNumberOfCols()))
+  auto rtype_enum = getParam<MooseEnum>("regression_type");
+  if (rtype_enum == "auto")
+    _rtype = _quad_sampler ? 0 : 1;
+  else
+    _rtype = rtype_enum == "integration" ? 0 : 1;
+
+  if (_rtype == 0 && _quad_sampler &&
+      (!_pvals.empty() || _pcols.size() != _sampler.getNumberOfCols()))
```
Comment (grmnptr) [inline modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C:59]: I think we should add an error if the number of polynomial terms is larger than the number of samples when we choose OLS.

File: modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md [hunk 5]
```diff
@@ -42,20 +42,40 @@ The weighting functions are defined by the probability density function of the p
...
+
+!plot scatter caption=Comparison between OLS regression and integration methods for computing polynomial chaos coefficients
```
Comment (grmnptr) [inline modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md:61]: Beautiful!

**Review by grmnptr — COMMENTED** [review]
Nice, just a few questions/comments.

File: modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md (hunk 5 above)
Comment (zachmprince) [reply modules/stochastic_tools/doc/content/source/surrogates/PolynomialChaos.md:61]: Thank you Peter!

File: modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C (hunk 3 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C:190]: Ended up adding it

**Comment by zachmprince** [comment]
> Do you get a covariance matrix for the coefficients from this?

@dschwen You definitely can, but it's a little more complicated with polynomial chaos since we are using orthogonal polynomials versus a simple linear regression. A more useful output of polynomial chaos is the computation of Sobol indices, which are similar to covariance.

**Review by grmnptr — APPROVED** [review]
